[BUG] 0.32.0 wheels ship py.typed but no .pyi stubs — breaks type checking of mlx.core downstream

### Summary

The macOS arm64 wheels for **0.32.0** ship `mlx/py.typed` but contain **no `.pyi` files at all**. The 0.31.2 wheels contain 8 (`mlx/core/__init__.pyi`, `cuda.pyi`, `distributed.pyi`, `fast.pyi`, …).

This is the worst of both worlds for downstream type checking: `py.typed` asserts the package is typed, so checkers trust it and do not fall back to inference or bundled third-party stubs — but there is nothing to resolve `mlx.core` against, since it is a compiled extension.

### Reproduction

No install needed — inspect the published wheels directly:

```python
import urllib.request, json, io, zipfile

for ver in ("0.31.2", "0.32.0"):
    d = json.load(urllib.request.urlopen(f"https://pypi.org/pypi/mlx/{ver}/json"))
    url = next(f["url"] for f in d["urls"]
               if f["filename"].endswith("cp312-cp312-macosx_14_0_arm64.whl"))
    names = zipfile.ZipFile(io.BytesIO(urllib.request.urlopen(url).read())).namelist()
    pyi = [n for n in names if n.endswith(".pyi")]
    print(ver, "pyi:", len(pyi), "py.typed:", "mlx/py.typed" in names, "files:", len(names))
```

```
0.31.2 pyi: 8 py.typed: True files: 47
0.32.0 pyi: 0 py.typed: True files: 39
```

Same result for `cp311`. (Checked on 2026-07-25; 0.32.0 is the latest release.)

### Downstream impact

Any type checker fails on every import. With `ty`:

```
error[unresolved-import]: Cannot resolve imported module `mlx.core`
 --> src/…/attention.py:3
  |
3 | import mlx.core as mx
```

Runtime is completely unaffected — this is packaging-only. In our project we had to constrain the **dev** environment to `mlx<0.32` to keep the type-check job green, which in turn blocks adopting any 0.32-only API (e.g. the `mx.fast.metal_kernel` math-mode option from #3728), since a `mlx>=0.32` runtime floor would make the dev extra uninstallable.

### Possible cause

Stub generation does not appear to have been removed deliberately — `setup.py` still runs it at the end of `build_ext.run()`:

```python
# Build type stubs.
build_temp = Path(self.build_temp) / ext.name
subprocess.run(
    ["cmake", "--install", build_temp, "--component", "core_stub"],
    check=True,
)
```

So the stubs are most likely generated but then dropped before the wheel is assembled. One hypothesis worth checking first: the two-stage release build (`MLX_BUILD_STAGE=1` for `mlx`, stage 2 for `mlx-metal`) combined with the CI rework in #3491, which moved `python setup.py clean --all` to run *after* `python -m build -w` and added `uv venv --clear` before the build. Something in that ordering could remove or hide the installed stubs relative to the directory that gets packaged. I have not bisected this — treat it as a starting point, not a diagnosis.

Prior art for the same class of regression: #2398 ("Fix including stubs in wheel") and #2942 ("Fix stubgen").

### Expected

0.32.x wheels ship `mlx/core/*.pyi` again, as 0.31.2 did. Failing that, dropping `mlx/py.typed` would at least let downstream checkers fall back gracefully instead of hard-failing — but restoring the stubs is clearly the better outcome.

A regression test in the release job — assert the built wheel contains `mlx/core/__init__.pyi` before upload — would prevent a third recurrence.

@zcbenz thanks for looking at this — happy to give you a concrete recipe.

One small clarification so CI doesn't end up green by accident: `uv check` verifies a project's metadata/lockfile rather than type stubs, so it will pass either way here. The checks below are aimed specifically at the `.pyi` files.

## The cheap gate: assert the stubs are in the wheel

This is the one that would actually have caught the regression, and it needs no type checker:

```bash
python -m zipfile -l dist/mlx-*.whl | grep -E 'mlx/core/.*\.pyi'
```

Or as a hard CI check right after the frontend wheel is built:

```bash
python - <<'EOF'
import glob, sys, zipfile
whl = glob.glob("dist/mlx-*.whl")[0]
names = zipfile.ZipFile(whl).namelist()
missing = [f for f in ("mlx/core/__init__.pyi", "mlx/core/fast.pyi",
                       "mlx/core/random.pyi", "mlx/core/linalg.pyi")
           if f not in names]
if missing:
    sys.exit(f"{whl}: missing stubs {missing}")
print(f"{whl}: stubs OK")
EOF
```

For reference, the frontend wheels contain (same platform tag, `cp312-macosx_15_0_arm64`):

| version | `mlx/py.typed` | `mlx/core/*.pyi` |
|---|---|---|
| 0.31.2 | yes | 8 files (`__init__`, `cuda`, `distributed`, `fast`, `fft`, `linalg`, `metal`, `random`) |
| 0.32.0 | yes | **0 files** |

## The end-to-end check: type-check a downstream snippet

If you want CI to verify that the shipped package is actually usable by a type checker (this is what breaks for downstream projects):

```bash
uv pip install mlx mypy
cat > stub_probe.py <<'EOF'
import mlx.core as mx

a: mx.array = mx.zeros((2, 3), dtype=mx.float32)
b: mx.array = mx.fast.rms_norm(a, mx.ones((3,)), 1e-5)
reveal_type(mx.zeros)
EOF
mypy stub_probe.py
```

Verified A/B in a clean venv (Python 3.12, macOS arm64), only the mlx version differs:

```
# mlx==0.32.0
stub_probe.py:1: error: Cannot find implementation or library stub for module named "mlx.core"  [import-not-found]
stub_probe.py:5: note: Revealed type is "Any"
Found 1 error in 1 file (checked 1 source file)

# mlx==0.31.2
stub_probe.py:5: note: Revealed type is "def (shape: int | typing.Sequence[int], dtype: mlx.core.Dtype | None =, *, stream: None | mlx.core.Stream | mlx.core.Device =) -> mlx.core.array"
Success: no issues found in 1 source file
```

Any checker works here (`pyright`, `ty`, …); mypy just has the least setup. Note that because `py.typed` is shipped, the missing stubs are *worse* than having no typing marker at all: downstream checkers trust the package and silently resolve everything through the extension module.

(`python -m mypy.stubtest mlx.core` is the tool for verifying stubs match the runtime, but it already reports a handful of pre-existing nits on 0.31.2, so I wouldn't make it a blocking gate — the two checks above are enough for this regression.)

## Where it likely regressed

The stub machinery itself is unchanged between 0.31.2 and 0.32.0 — `python/src/CMakeLists.txt` is byte-identical, and `setup.py` still has both `nanobind_add_stub`/`--component core_stub` and `package_data = {"mlx.core": ["*.pyi"]}`.

What changed is how release wheels are built. In `.github/actions/build-wheel/action.yml` the frontend step now runs:

```
python setup.py clean --all
MLX_BUILD_STAGE=1 python -m build -w
```

The stubs are generated in `CMakeBuild.run()` (i.e. during `build_ext`), which setuptools runs *after* `build_py` — and `build_py` is what collects `package_data` from `python/mlx/core/*.pyi`. So on a tree that already has stubs from a previous build the files get picked up, but on a freshly cleaned tree (`clean --all` + a fresh CI checkout) `build_py` runs before they exist and the wheel ships without them. That would explain why local builds look fine while the released wheels don't.

Haven't tested a fix, so treat that as a hypothesis — but if it's right, generating the stubs before `build_py` (or adding the `.pyi` files via a `build_py`/`bdist_wheel` hook after `build_ext`) plus the zip assertion above should close it for good.

Happy to open a PR if that's useful.

--- a/.github/actions/build-macos/action.yml
+++ b/.github/actions/build-macos/action.yml
@@ -17,7 +17,7 @@ runs:
       shell: bash
       run: |
         echo "::group::Install dependencies"
-        uv pip install 'build<=1.4.2' setuptools
+        uv pip install build setuptools
         echo "::endgroup::"
 
     - name: Build wheel

--- a/.github/actions/build-macos/action.yml
+++ b/.github/actions/build-macos/action.yml
@@ -15,7 +15,7 @@ runs:
   steps:
     - name: Install dependencies
       shell: bash
-      run: uv pip install 'build<=1.4.2' setuptools
+      run: uv pip install build setuptools
 
     - name: Build wheel
       shell: bash

--- a/.github/actions/build-wheel/action.yml
+++ b/.github/actions/build-wheel/action.yml
@@ -35,7 +35,7 @@ runs:
       shell: bash
       run: |
         echo "::group::Install dependencies"
-        uv pip install 'build<=1.4.2' setuptools
+        uv pip install build setuptools
         if ${{ runner.os == 'Linux' }} ; then
           uv pip install auditwheel patchelf
         fi
@@ -50,8 +50,7 @@ runs:
         MACOSX_DEPLOYMENT_TARGET: ${{ inputs.macos-target }}
       run: |
         echo "::group::Build frontend package"
-        python setup.py clean --all
-        MLX_BUILD_STAGE=1 python -m build -w
+        MLX_BUILD_FRONTEND_PACKAGE=1 python -m build -w
         echo "::endgroup::"
 
     - name: Post-process frontend package
@@ -77,8 +76,8 @@ runs:
         MACOSX_DEPLOYMENT_TARGET: ${{ inputs.macos-target }}
       run: |
         echo "::group::Build backend package"
+        MLX_BUILD_BACKEND_PACKAGE=1 python -m build -w
         python setup.py clean --all
-        MLX_BUILD_STAGE=2 python -m build -w
         echo "::endgroup::"
 
     - name: Post-process backend package

--- a/.github/actions/build-wheel/action.yml
+++ b/.github/actions/build-wheel/action.yml
@@ -34,7 +34,7 @@ runs:
     - name: Install dependencies
       shell: bash
       run: |
-        uv pip install 'build<=1.4.2' setuptools
+        uv pip install build setuptools
         if ${{ runner.os == 'Linux' }} ; then
           uv pip install auditwheel patchelf
         fi
@@ -48,7 +48,7 @@ runs:
         MACOSX_DEPLOYMENT_TARGET: ${{ inputs.macos-target }}
       run: |
         python setup.py clean --all
-        MLX_BUILD_STAGE=1 python -m build -w
+        MLX_BUILD_FRONTEND_PACKAGE=1 python -m build -w
 
     - name: Post-process frontend package
       if: inputs.build-frontend == 'true'
@@ -71,7 +71,7 @@ runs:
         MACOSX_DEPLOYMENT_TARGET: ${{ inputs.macos-target }}
       run: |
         python setup.py clean --all
-        MLX_BUILD_STAGE=2 python -m build -w
+        MLX_BUILD_BACKEND_PACKAGE=1 python -m build -w
 
     - name: Post-process backend package
       if: inputs.build-backend == 'true'

--- a/.github/actions/setup/action.yml
+++ b/.github/actions/setup/action.yml
@@ -42,7 +42,7 @@ runs:
         echo "::group::Install common dependencies"
         sudo apt-get update
         sudo apt-get install -y --no-install-recommends \
-            gdb g++ ninja-build zip \
+            gdb g++ ninja-build unzip \
             libblas-dev liblapack-dev liblapacke-dev \
             openmpi-bin openmpi-common libopenmpi-dev
         echo "::endgroup::"
@@ -53,6 +53,7 @@ runs:
       run: |
         echo "::group::Install macOS dependencies"
         brew update
+        brew trust aws/tap # suppress warning in github actions
         brew install openmpi
         xcodebuild -showComponent MetalToolchain
         sysctl -a | grep machdep.cpu

--- a/.github/actions/setup/action.yml
+++ b/.github/actions/setup/action.yml
@@ -41,7 +41,7 @@ runs:
       run: |
         sudo apt-get update
         sudo apt-get install -y --no-install-recommends \
-            gdb g++ ninja-build zip \
+            gdb g++ ninja-build unzip \
             libblas-dev liblapack-dev liblapacke-dev \
             openmpi-bin openmpi-common libopenmpi-dev
 
@@ -50,6 +50,7 @@ runs:
       shell: bash
       run: |
         brew update
+        brew trust aws/tap # suppress warning in github actions
         brew install openmpi
         xcodebuild -showComponent MetalToolchain
         sysctl -a | grep machdep.cpu

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -86,6 +86,7 @@ jobs:
         with:
           cmake-args: ${{ steps.setup.outputs.cmake-args }}
           build-backend: false
+      - run: unzip -l wheelhouse/*.whl
       - uses: actions/upload-artifact@v7
         with:
           name: frontend-${{ runner.os }}-${{ runner.arch }}-py${{ matrix.python-version }}
@@ -127,6 +128,7 @@ jobs:
         with:
           cmake-args: ${{ steps.setup.outputs.cmake-args }}
           build-frontend: false
+      - run: unzip -l wheelhouse/*.whl
       - uses: actions/upload-artifact@v7
         with:
           name: backend-${{ matrix.toolkit }}-${{ runner.os }}-${{ runner.arch }}
@@ -168,6 +170,8 @@ jobs:
           macos-target: '26.2'
           cmake-args: ${{ steps.setup.outputs.cmake-args }}
           build-backend: ${{ matrix.python-version == '3.10' }}
+      - name: Display content of the wheels
+        run: for WHEEL in wheelhouse/*.whl; do unzip -l $WHEEL; echo; done
       - name: Upload frontend packages
         uses: actions/upload-artifact@v7
         with:

--- a/setup.py
+++ b/setup.py
@@ -53,7 +53,22 @@ def get_version():
     return version
 
 
-build_stage = int(os.environ.get("MLX_BUILD_STAGE", 0))
+# Release builds for PyPi are separated into 2 packages:
+#
+# Frontend package:
+#  - Triggered with `MLX_BUILD_FRONTEND_PACKAGE=1`
+#  - Include everything except backend-specific binaries (e.g. libmlx.so, mlx.metallib, etc)
+#  - Wheel has Python ABI and platform tags
+#  - Wheel should be built for the cross-product of python version and platforms
+#  - Package name is "mlx" and it depends on backend packages (e.g. mlx-metal, mlx-cuda)
+# Backend package:
+#  - Triggered with `MLX_BUILD_BACKEND_PACKAGE=1`
+#  - Include headers and backend binaries.
+#  - Wheel has only platform tags
+#  - Wheel should be built only for different platforms
+#  - Package name is back-end specific, e.g mlx-metal, mlx-cuda
+build_frontend = int(os.environ.get("MLX_BUILD_FRONTEND_PACKAGE", 0))
+build_backend = int(os.environ.get("MLX_BUILD_BACKEND_PACKAGE", 0))
 build_macos = platform.system() == "Darwin"
 build_cuda = "MLX_BUILD_CUDA=ON" in os.environ.get("CMAKE_ARGS", "")
 
@@ -77,9 +92,7 @@ class CMakeBuild(build_ext):
             self.build_temp = os.path.dirname(self.build_temp)
 
     def build_extension(self, ext: CMakeExtension) -> None:
-        # Must be in this form due to bug in .resolve() only fixed in Python 3.10+
-        ext_fullpath = Path.cwd() / self.get_ext_fullpath(ext.name)  # type: ignore[no-untyped-call]
-        extdir = ext_fullpath.parent.resolve()
+        extdir = self._get_ext_dir(ext)
 
         debug = int(os.environ.get("DEBUG", 0)) if self.debug is None else self.debug
         cfg = "Debug" if debug else "Release"
@@ -88,17 +101,9 @@ class CMakeBuild(build_ext):
         if not build_temp.exists():
             build_temp.mkdir(parents=True)
 
-        install_prefix = extdir
-        pybind_out_dir = extdir
-        if build_stage == 1:
-            # Don't include MLX libraries in the wheel
-            install_prefix = build_temp
-        elif build_stage == 2:
-            # Don't include Python bindings in the wheel
-            pybind_out_dir = build_temp
         cmake_args = [
-            f"-DCMAKE_INSTALL_PREFIX={install_prefix}",
-            f"-DMLX_PYTHON_BINDINGS_OUTPUT_DIRECTORY={pybind_out_dir}",
+            f"-DCMAKE_INSTALL_PREFIX={extdir}",
+            f"-DMLX_PYTHON_BINDINGS_OUTPUT_DIRECTORY={extdir}",
             f"-DCMAKE_BUILD_TYPE={cfg}",
             f"-DPython_EXECUTABLE={sys.executable}",
             "-DMLX_BUILD_PYTHON_BINDINGS=ON",
@@ -113,8 +118,7 @@ class CMakeBuild(build_ext):
         if "CMAKE_ARGS" in os.environ:
             cmake_args += [item for item in os.environ["CMAKE_ARGS"].split(" ") if item]
 
-        # For release wheel force building for all supported arches.
-        if build_stage == 2 and build_cuda:
+        if build_backend and build_cuda:
             # Last arch is always real and virtual for forward-compatibility
             cuda_archs = [
                 "75-real",
@@ -186,15 +190,49 @@ class CMakeBuild(build_ext):
             ["cmake", "--install", build_temp, "--component", "core_stub"],
             check=True,
         )
+        # Copy the type stubs to extdir so they are included in wheels.
+        stubs_dir = Path("python/mlx/core")
+        if stubs_dir.exists():
+            extdir = self._get_ext_dir(ext)
+            self.copy_tree(stubs_dir, extdir / "core")
+
+    def _get_ext_dir(self, ext):
+        # Must be in this form due to bug in .resolve() only fixed in Python 3.10+
+        ext_fullpath = Path.cwd() / self.get_ext_fullpath(ext.name)  # type: ignore[no-untyped-call]
+        return ext_fullpath.parent.resolve()
 
 
 class MLXBdistWheel(bdist_wheel):
     def get_tag(self) -> tuple[str, str, str]:
         impl, abi, plat_name = super().get_tag()
-        if build_stage == 2:
+        if build_backend:
             impl = self.python_tag
             abi = "none"
         return (impl, abi, plat_name)
+
+    def write_wheelfile(self, *args, **kwargs) -> None:
+        super().write_wheelfile(*args, **kwargs)
+
+        mlx_dir = Path(self.bdist_dir, "mlx")
+
+        def is_backend_file(file):
+            if file.is_relative_to(Path(mlx_dir, "lib")):
+                return True
+            if file.is_relative_to(Path(mlx_dir, "include")):
+                return True
+            if file.is_relative_to(Path(mlx_dir, "share")):
+                return True
+            if file.suffix == ".dll":
+                return True
+            return False
+
+        if build_frontend or build_backend:
+            for file in Path(self.bdist_dir).rglob("*"):
+                if not file.is_relative_to(mlx_dir) or not file.is_file():
+                    continue
+                bf = is_backend_file(file)
+                if (build_frontend and bf) or (build_backend and not bf):
+                    file.unlink()
 
 
 # Read the content of README.md
@@ -260,24 +298,8 @@ if __name__ == "__main__":
     }
     install_requires = []
 
-    # Release builds for PyPi are in two stages.
-    # Each stage should be run from a clean build:
-    #   python setup.py clean --all
-    #
-    # Stage 1:
-    #  - Triggered with `MLX_BUILD_STAGE=1`
-    #  - Include everything except backend-specific binaries (e.g. libmlx.so, mlx.metallib, etc)
-    #  - Wheel has Python ABI and platform tags
-    #  - Wheel should be built for the cross-product of python version and platforms
-    #  - Package name is mlx and it depends on subpackage in stage 2 (e.g. mlx-metal)
-    # Stage 2:
-    #  - Triggered with `MLX_BUILD_STAGE=2`
-    #  - Includes only backend-specific binaries (e.g. libmlx.so, mlx.metallib, etc)
-    #  - Wheel has only platform tags
-    #  - Wheel should be built only for different platforms
-    #  - Package name is back-end specific, e.g mlx-metal
-    if build_stage != 2:
-        if build_stage == 1:
+    if not build_backend:
+        if build_frontend:
             install_requires.append(
                 f'mlx-metal=={version}; platform_system == "Darwin"'
             )

